enum.c

```diff
@@ -579,6 +579,68 @@ enum_filter_map(VALUE obj)
     return ary;
 }
 
+static VALUE
+join_map_i(RB_BLOCK_CALL_FUNC_ARGLIST(i, args))
+{
+    struct MEMO *memo = MEMO_CAST(args);
+    VALUE result = memo->v1;
+    VALUE sep = memo->v2;
+    VALUE val;
+
+    ENUM_WANT_SVALUE();
+
+    val = enum_yield(argc, i);
+
+    if (memo->u3.cnt > 0 && !NIL_P(sep)) {
+        rb_str_buf_append(result, sep);
+    }
+    rb_str_buf_append(result, rb_obj_as_string(val));
+    memo->u3.cnt++;
+
+    return Qnil;
+}
+
+/*
+ * call-seq:
+ *   join_map(separator = "") {|element| ... } -> string
+ *   join_map(separator = "") -> enumerator
+ *
+ * Returns a string created by converting each element to a string via the block,
+ * then joining the results with the given separator.
+ *
+ * With a block given, calls the block with successive elements;
+ * converts each block return value to a string and joins them:
+ *
+ *   users = [{name: "Alice"}, {name: "Bob"}, {name: "Charlie"}]
+ *   users.join_map(", ") { |user| user[:name] } # => "Alice, Bob, Charlie"
+ *   (1..5).join_map("-") { |n| n * 2 }           # => "2-4-6-8-10"
+ *
+ * When no block given, returns an Enumerator.
+ *
+ */
+static VALUE
+enum_join_map(int argc, VALUE *argv, VALUE obj)
+{
+    VALUE sep, result;
+    struct MEMO *memo;
+
+    rb_scan_args(argc, argv, "01", &sep);
+    RETURN_SIZED_ENUMERATOR(obj, argc, argv, enum_size);
+
+    if (NIL_P(sep)) {
+        sep = rb_str_new_cstr("");
+    }
+    else {
```

Comment:
`join_map_i` checks if `!NIL_P(sep)`.

```suggestion
    if (!NIL_P(sep)) {
```